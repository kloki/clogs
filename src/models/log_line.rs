use std::collections::BTreeMap;

use chrono::{DateTime, Utc};
use colorize::AnsiColor;
use serde::{Deserialize, Deserializer};
use serde_json::Value;

use super::log_level::LogLevel;

#[derive(Deserialize)]
pub struct LogLine {
    pub timestamp: DateTime<Utc>,
    pub level: LogLevel,
    #[serde(deserialize_with = "deserialize_fields")]
    pub fields: BTreeMap<String, String>,
}

impl LogLine {
    fn timestamp_clog(&self) -> String {
        format!("[{}]", self.timestamp.format("%H:%M:%S").to_string()).grey()
    }

    fn label_clog(&self) -> String {
        match self.fields.get("label") {
            Some(content) => format!("[{}] ", content.clone().cyan()),
            None => "".to_string(),
        }
    }
    fn fields_clog(&self) -> String {
        self.fields
            .iter()
            .filter(|(k, _)| k != &"message" && k != &"label")
            .map(|(k, v)| {
                format!(
                    "\n           {}:{}",
                    format!("{:<12}", k).cyan(),
                    v.to_string().grey()
                )
            })
            .collect::<Vec<_>>()
            .join("")
    }
    pub fn to_single_clog(&self) -> String {
        format!(
            "{} {} {}{}",
            self.timestamp_clog(),
            self.level.to_clog(),
            self.label_clog(),
            self.fields.get("message").cloned().unwrap_or_default(),
        )
    }

    pub fn to_multi_clog(&self) -> String {
        format!("{}{}", self.to_single_clog(), self.fields_clog())
    }
}

fn deserialize_fields<'de, D>(deserializer: D) -> Result<BTreeMap<String, String>, D::Error>
where
    D: Deserializer<'de>,
{
    let raw_fields: BTreeMap<String, Value> = BTreeMap::deserialize(deserializer)?;
    let fields = raw_fields
        .into_iter()
        .map(|(k, v)| (k, v.to_string().trim_matches('"').to_string()))
        .collect();
    Ok(fields)
}
