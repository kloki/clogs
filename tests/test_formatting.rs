use std::fs;

use clogs::models::log_line::LogLine;

#[test]
fn test_single_clog() {
    let input = fs::read_to_string("tests/input/test.log").expect("Unable to read file");
    let result = input
        .lines()
        .map(|line| {
            serde_json::from_str::<LogLine>(&line)
                .unwrap()
                .to_single_clog()
        })
        .collect::<Vec<_>>()
        .join("\n");
    println!("{}", result);
    insta::assert_yaml_snapshot!(result);
}

#[test]
fn test_multi_clog() {
    let input = fs::read_to_string("tests/input/test.log").expect("Unable to read file");
    let result = input
        .lines()
        .map(|line| {
            serde_json::from_str::<LogLine>(&line)
                .unwrap()
                .to_multi_clog()
        })
        .collect::<Vec<_>>()
        .join("\n");
    println!("{}", result);
    insta::assert_yaml_snapshot!(result);
}
