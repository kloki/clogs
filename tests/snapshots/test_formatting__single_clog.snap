---
source: tests/test_formatting.rs
expression: result
---
"\u001b[37m[12:34:56]\u001b[0;39;49m \u001b[34mINFO \u001b[0;39;49m User successfully logged in.\n\u001b[37m[12:35:01]\u001b[0;39;49m \u001b[33mWARN \u001b[0;39;49m Login attempt failed due to invalid password.\n\u001b[37m[12:35:10]\u001b[0;39;49m \u001b[31mERROR\u001b[0;39;49m Database connection failed.\n\u001b[37m[12:36:00]\u001b[0;39;49m \u001b[35mTRACE\u001b[0;39;49m Background job processing started.\n\u001b[37m[12:36:30]\u001b[0;39;49m \u001b[32mDEBUG\u001b[0;39;49m User profile updated.\n\u001b[37m[12:36:30]\u001b[0;39;49m \u001b[41;30mCRITICAL\u001b[0;39;49m Missing funds\n\u001b[37m[12:35:10]\u001b[0;39;49m \u001b[35mTRACE\u001b[0;39;49m [\u001b[36mmetrics\u001b[0;39;49m] \n\u001b[37m[12:35:10]\u001b[0;39;49m \u001b[35mTRACE\u001b[0;39;49m [\u001b[36mmetrics\u001b[0;39;49m] very high temp!"
